Homework1

ZibyLC opened the pull request:
Проект создан на основе Cookiecutter DataScience. 

Основные решения:
1. Шаблон Cookiecutter Data Science
2. Makefile для работы с make
3. Использовано логирование
4. Быстрый старт: git clone, cd ml_project, make train, make predict

Критерии (указаны максимальные баллы, по каждому критерию ревьюэр может поставить баллы частично):
- [x] В описании к пулл реквесту описаны основные "архитектурные" и тактические решения, которые сделаны в вашей работе. В общем, описание того, что именно вы сделали и для чего, чтобы вашим ревьюерам было легче понять ваш код (1 балл)
- [x] В пулл-реквесте проведена самооценка, распишите по каждому пункту выполнен ли критерий или нет и на сколько баллов(частично или полностью) (1 балл)
- [x] Выполнено EDA, закоммитьте ноутбук в папку с ноутбуками (1 балл) Вы так же можете построить в ноутбуке прототип(если это вписывается в ваш стиль работы) -  проведён анализ данных в notebooks/eda.ipynb
- [x] Можете использовать не ноутбук, а скрипт, который сгенерит отчет, закоммитьте и скрипт и отчет (за это + 1 балл) - make eda генерирует отчёт с помощью pandas profiling  и сохраняет его в reports/ в виде html и json. 
- [x] Написана функция/класс для тренировки модели, вызов оформлен как утилита командной строки, записана в readme инструкцию по запуску (3 балла) - в readme описаны команды make data, make train
- [x] Написана функция/класс predict (вызов оформлен как утилита командной строки), которая примет на вход артефакт/ы от обучения, тестовую выборку (без меток) и запишет предикт по заданному пути, инструкция по вызову записана в readme (3 балла) - make predict
- [x] Проект имеет модульную структуру (2 балла)
- [x] Использованы логгеры (2 балла) - логов мало, только самое важное, но и действий немного
- [ ] Написаны тесты на отдельные модули и на прогон обучения и predict (3 балла) - не сделал
- [ ] Для тестов генерируются синтетические данные, приближенные к реальным (2 балла) - можно посмотреть на библиотеки https://faker.readthedocs.io/, https://feature-forge.readthedocs.io/en/latest/, можно просто руками посоздавать данных, собственноручно написанными функциями. - для трансформеров написана тесты, где используется fake
- [ ] Обучение модели конфигурируется с помощью конфигов в json или yaml, закоммитьте как минимум 2 корректные конфигурации, с помощью которых можно обучить модель (разные модели, стратегии split, preprocessing) (3 балла) - два примитивных конфета для двух моделей: logreg.yaml и tree.yaml
- [ ] Используются датаклассы для сущностей из конфига, а не голые dict (2 балла) - используется omegadict
- [ ] Напишите кастомный трансформер и протестируйте его (3 балла) https://towardsdatascience.com/pipelines-custom-transformers-in-scikit-learn-the-step-by-step-guide-with-python-code-4a7d9b068156 - сделано два кастомных трансформера MultiOneHotEncoder и FeatureSelection - для MultiOneHotEncoder написан простой тест (отчасти бесполезный)
- [x] В проекте зафиксированы все зависимости (1 балл) -  pipenv -> requirements.txt
- [ ] Настроен CI для прогона тестов, линтера на основе github actions (3 балла). Пример с пары: https://github.com/demo-ml-cicd/ml-python-package
- [ ] Используйте hydra для конфигурирования (https://hydra.cc/docs/intro/) - 3 балла. - src/configs

ZibyLC committed:
first real commit

ZibyLC committed:
first commit

ZibyLC committed:
changed console utilites & makefile

ZibyLC committed:
eda report added

ZibyLC committed:
train and predict model added

ZibyLC committed:
Makefile hotfix

ZibyLC committed:
readme filled and minor bugs fixed

+Fixed loggers in predict
+Removed target column from make_dataset according to homework description

SVizor42 commented on ml_project/src/models/predict_model.py:
```diff
@@ -0,0 +1,50 @@
+from pathlib import Path
+from dotenv import find_dotenv, load_dotenv
+import pandas as pd
+from sklearn.linear_model import LogisticRegression
```
Импортируется, но не используется

SVizor42 commented on ml_project/src/models/predict_model.py:
```diff
@@ -0,0 +1,50 @@
+    logger = logging.getLogger(__name__)
+    logger.info('predict with model')
+
+    df = pd.read_csv(data_path)
```
В продакшене следует давать более осмысленные имена переменным

SVizor42 commented on ml_project/src/models/predict_model.py:
```diff
@@ -0,0 +1,50 @@
+@click.argument('data_path', type=click.Path(), default='data/processed/test.csv')
+@click.argument('model_path', type=click.Path(), default='models/estimator.pkl')
+@click.argument('result_path', type=click.Path(), default='data/processed/predict_proba.csv')
+def main(data_path, model_path, result_path):
```
Лучше не использовать безликое `main`. + нужно обратить внимание на Typing

SVizor42 commented on ml_project/src/models/predict_model.py:
```diff
@@ -0,0 +1,50 @@
+
+    # find .env automagically by walking up directories until it's found, then
+    # load up the .env entries as environment variables
+    load_dotenv(find_dotenv())
```
Для чего здесь? Используются какие-то secrets?

ZibyLC replied:
Здравствуйте, это было в шаблоне. По сути про запас, чтобы можно было добавить параметры окружения. Наверно стоило подчистить перед коммитом

SVizor42 commented on ml_project/src/data/make_dataset.py:
```diff
@@ -0,0 +1,63 @@
+    logger.info(f'preprocessing successfully completed')
+
+    if make_split:
+        train_df, test_df = train_test_split(df, test_size=0.2, stratify=df['target'], random_state=103)
```
Следует избегать использования MagicNumbers в проде. Если нет конфига, можно вынести в раздел констант.

SVizor42 commented on ml_project/src/models/train_model.py:
```diff
@@ -0,0 +1,47 @@
+    logger.info(f'df splitted on X, y according to features/target list')
+    estimator.fit(X, y)
+    logger.info(f'model successfully fitted')
+    dump(estimator, model_path)
```
Прежде модель желательно провалидировать и посчитать метрики

SVizor42 reviewed (changes requested):
Отдельный плюс за использование Makefile.
Что касается замечаний, то см. комментарии. Некоторые из них носят общий характер.
Сюда же добавлю:
1. Не стоит коммитить пустые папки и файлы.
2. Есть вопросы к модульности - всего 3 основных модуля.
3. Желательно еще раз прогнать код через flake или другой линтер.

ZibyLC committed:
commit to meet pull-request supervisor's commentaries

1. names of varriables changed to something more meaningful and informative
2. Typing added. "->" is not used, because functions don't return anything
3. Useless imports deleted
4. load_dotenv - removed
5. MagicNumbers transfered out from functions to global constants
6. In train_model - added cross_validation and estimation on roc_auc metric
7. Most of linter warnings fixed. Some varriables like "X" are saved corresponding to sklearn codestyle
8. empty folders and useless files removed from project structure

ZibyLC commented:
Благодарю за обратную свзяь. Внимательно ознакомился со всеми замечаниями и пофиксил проект в последнем коммите. Отправляю в мердж

SVizor42 commented:
**Сумма: 15 * 0.6 = 9 баллов**

Хорошо. Только мерджи в следующий раз после аппрува, пжлста.
